File: frontend/src/types/marketplace.ts
// Marketplace-related TypeScript types

export interface MarketStory {
  id: number;
  title: string;
  author: string;
  author_id: string;
  content?: string;
  ai_summary: string;
  ai_genres: string[];
  average_rating: number;
  rating_count: number;
  total_downloads: number;
  total_donated_credits: number;
  published_at: string;
  is_staff_pick: boolean;
  user_rating?: number; // User's own rating if authenticated
}

export interface MarketStoryCard {
  id: number;
  title: string;
  author: string;
  ai_summary: string;
  ai_genres: string[];
  average_rating: number;
  rating_count: number;
  total_downloads: number;
  is_staff_pick: boolean;
}

export interface MarketStoriesResponse {
  stories: MarketStoryCard[];
  pagination: {
    page: number;
    per_page: number;
    total: number;
    pages: number;
  };
}

export interface PublishStoryData {
  title: string;
  allow_ai: boolean;
  terms_accepted: boolean;
}

export interface RatingData {
  rating: number;
}

export interface DonationData {
  credits: number;
}

export interface MarketplaceFilters {
  search?: string;
  genre?: string;
  min_rating?: number;
  sort_by?: 'published_at' | 'rating' | 'downloads' | 'donations' | 'title';
  sort_order?: 'asc' | 'desc';
}

export interface MarketplaceSectionResponse {
  stories: MarketStoryCard[];
}

File: backend/models/marketplace.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class PublishStoryRequest(BaseModel):
    title: str = Field(..., min_length=1)
    allow_ai: bool = True
    terms_accepted: bool = Field(..., description="Must accept terms and conditions")

class PublishStoryResponse(BaseModel):
    success: bool
    market_story_id: int
    message: str

class MarketStoryListItem(BaseModel):
    id: int
    title: str
    author: str
    ai_summary: Optional[str] = None
    ai_genres: List[str] = []
    average_rating: float
    rating_count: int
    total_downloads: int
    total_donated_credits: int
    published_at: str
    is_staff_pick: bool
    image_uri: Optional[str] = None

class PaginationInfo(BaseModel):
    page: int
    per_page: int
    total: int
    pages: int

class MarketStoriesResponse(BaseModel):
    stories: List[MarketStoryListItem]
    pagination: PaginationInfo

class MarketStoryDetail(BaseModel):
    id: int
    title: str
    author: str
    content: str
    ai_summary: Optional[str] = None
    ai_genres: List[str] = []
    average_rating: float
    rating_count: int
    total_downloads: int
    total_donated_credits: int
    published_at: str
    is_staff_pick: bool
    author_id: str
    image_uri: Optional[str] = None
    scenario_json: Optional[str] = None
    user_rating: Optional[int] = None

class DownloadResponse(BaseModel):
    success: bool

class RatingRequest(BaseModel):
    rating: int = Field(..., ge=1, le=5)

class RatingResponse(BaseModel):
    success: bool
    average_rating: float
    rating_count: int

class DonationRequest(BaseModel):
    credits: int = Field(..., ge=1)

class DonationResponse(BaseModel):
    success: bool
    message: str

class StoriesListResponse(BaseModel):
    stories: List[MarketStoryListItem]

class Genre(BaseModel):
    name: str
    count: int

class GenresResponse(BaseModel):
    genres: List[Genre]

class UserCreditsResponse(BaseModel):
    credits: int
    cached: bool

class ClearCacheResponse(BaseModel):
    message: str